PageSpeed_Api.php

```diff
@@ -1,128 +1,1898 @@
+			if ( is_wp_error( $response ) ) {
+				return array(
+					'error' => array(
+						'message' => $response->get_error_message()
+					),
+				);
+			} elseif ( isset( $response['response']['code'] ) && 200 !== $response['response']['code'] ) {
+				return array(
+					'error' => array(
+						'code'    => $response['response']['code'],
+						'message' => $response['response']['message']
+					),
+				);
+			}
+		}
 
+		$access_token = json_decode( $this->client->getAccessToken() );
 
-	public function get_page_score( $url ) {
-		$json = $this->_request( array(
-				'url' => $url,
-				'category' => 'performance',
-				'strategy' => 'desktop'
-			) );
+		$request = Util_Environment::url_format(
+			W3TC_PAGESPEED_API_URL,
+			array_merge(
+				$query,
+				array(
+					'quotaUser'    => Util_Http::generate_site_id(),
+					'access_token' => $access_token->access_token,
+				)
+			)
+		);
+		
+		// Attempt the request up to 4 times with an increasing delay between each attempt
+		$attempts = 0;
+		while ( ++$attempts ) {
+			try {
+				$response = wp_remote_get(
+					$request,
+					array(
+						'timeout' => 60,
+					)
+				);
+				if ( ! is_wp_error( $response ) && 200 === $response['response']['code'] ) {
+					break;
+				}
+			} catch ( \Exception $e ) {
+				if ( $attempts <= 4 ) {
+					sleep( $attempts * 2 );
+				} else {
+					return array(
+						'error' => array(
+							'code'    => 500,
+							'message' => $e->getMessage(),
+						),
+					);
+				}
+			}
+		};
 
-		if ( !$json ) {
-			return null;
+		if ( is_wp_error( $response ) ) {
+			return wp_json_encode(
+				array(
+					'error' => array(
+						'message' => $response->get_error_message(),
+					),
+				)
+			);
+		} elseif ( isset( $response['response']['code'] ) && 200 !== $response['response']['code'] ) {
+			return wp_json_encode(
+				array(
+					'error' => array(
+						'code'    => $response['response']['code'],
+						'message' => $response['response']['message'],
+					),
+				)
+			);
 		}
 
-		$data = array();
-		try {
-			$data = json_decode( $json, true );
-		} catch ( \Exception $e ) {
+		return json_decode( wp_remote_retrieve_body( $response ), true );
+	}
+
+	/**
+	 * Recursively get value series of key decendents
+	 *
+	 * @param array $data PageSpeed data.
+	 * @param array $elements Elements to get values of.
+	 *
+	 * @return object
+	 */
+	public function v( $data, $elements ) {
+		if ( empty( $elements ) ) {
+			return $data;
+		}
+
+		$key = array_shift( $elements );
+		if ( ! isset( $data[ $key ] ) ) {
+			return null;
 		}
 
-		return $this->v( $data, array( 'lighthouseResult', 'categories', 'performance', 'score' ) );
+		return $this->v( $data[ $key ], $elements );
 	}
 
+	/**
+	 * Recursively get value series of key decendents
+	 *
+	 * @param array $data PageSpeed data.
+	 * @param array $elements Elements to get values of.
+	 *
+	 * @return object
+	 */
+	public function handle_( $data, $elements ) {
```

Comment:
Why does the function end with a `_`?

What does a function named `handle` actually do?